File: src/core/tokenBucket.ts
export interface TokenBucketOptions {
  capacity: number;
  refillRatePerSecond: number;
  initialTokens?: number;
  lastRefillTimestamp?: number;
}

export class TokenBucket {
  public readonly capacity: number;
  public readonly refillRatePerSecond: number;
  public tokens: number;
  public lastRefillTimestamp: number;

  constructor(options: TokenBucketOptions) {
    this.capacity = options.capacity;
    this.refillRatePerSecond = options.refillRatePerSecond;
    this.tokens = options.initialTokens !== undefined ? options.initialTokens : options.capacity;
    this.lastRefillTimestamp = options.lastRefillTimestamp !== undefined ? options.lastRefillTimestamp : 0;
  }

  /**
   * Attempts to consume 1 token from the bucket.
   * @param now The current timestamp in milliseconds.
   */
  public tryConsume(now: number): { allowed: boolean; remaining: number; resetAt: number } {
    // Lazy initialize the lastRefillTimestamp if it was not provided in construction
    if (this.lastRefillTimestamp === 0) {
      this.lastRefillTimestamp = now;
    }

    const elapsedMs = Math.max(0, now - this.lastRefillTimestamp);
    const elapsedSeconds = elapsedMs / 1000;

    // Refill the bucket based on elapsed wall-clock time
    this.tokens = Math.min(
      this.capacity,
      this.tokens + elapsedSeconds * this.refillRatePerSecond
    );
    this.lastRefillTimestamp = now;

    let allowed = false;
    if (this.tokens >= 1) {
      this.tokens -= 1;
      allowed = true;
    }

    // Return the floored count of remaining tokens
    const remaining = Math.floor(this.tokens);

    // Calculate when the bucket will be completely full again (at capacity)
    const missingTokens = this.capacity - this.tokens;
    const secondsToRefill = missingTokens / this.refillRatePerSecond;
    const resetAt = Math.ceil(now + secondsToRefill * 1000);

    return {
      allowed,
      remaining,
      resetAt,
    };
  }
}

File: src/services/luaScripts/tokenBucket.lua
-- KEYS[1] = bucket:{clientKey}
-- ARGV[1] = capacity (number)
-- ARGV[2] = refillRatePerSecond (number)
-- ARGV[3] = now (milliseconds)
-- ARGV[4] = requested tokens (number, default: 1)
--
-- Replicates the logic from TokenBucket.tryConsume (src/core/tokenBucket.ts)
-- Explicitly cross-referenced: if the refill formula or consumption logic
-- changes in tokenBucket.ts, it must be synchronized here as well.

local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refillRate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local requested = tonumber(ARGV[4]) or 1

-- Fetch existing state from Redis
local state = redis.call('HMGET', key, 'tokens', 'lastRefillTimestamp')
local tokens = tonumber(state[1])
local lastRefillTimestamp = tonumber(state[2])

if not tokens or not lastRefillTimestamp then
  -- Fresh initialization
  tokens = capacity
  lastRefillTimestamp = now
else
  -- Calculate elapsed time and refill
  local elapsedMs = math.max(0, now - lastRefillTimestamp)
  local elapsedSeconds = elapsedMs / 1000.0
  tokens = math.min(capacity, tokens + (elapsedSeconds * refillRate))
  lastRefillTimestamp = now
end

local allowed = 0
if tokens >= requested then
  tokens = tokens - requested
  allowed = 1
end

-- Save updated state back to Redis
redis.call('HMSET', key, 'tokens', tostring(tokens), 'lastRefillTimestamp', tostring(lastRefillTimestamp))

-- Calculate resetAt (when the bucket will be completely full again)
local missingTokens = capacity - tokens
local secondsToRefill = missingTokens / refillRate
local resetAt = math.ceil(now + (secondsToRefill * 1000.0))

-- Return allowed (1/0), remaining (floored tokens), resetAt
-- All returned values are integers in Lua and safely cast to integers/JS numbers.
return { allowed, math.floor(tokens), resetAt }
